04.bowling/bowling.rb

```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+score = ARGV[0]
+
+score_strings = score.split(',')
+
+scores = []
+score_strings.each do |n|
+  if n == 'X'
+    scores << 10
+    scores << 0
+  else
+    scores << n.to_i
+  end
+end
+
+frames = []
+scores.each_slice(2) do |f|
+  frames << f
+end
+
+frame_with_number = {}
+frames.each.with_index(1) { |fp, i| frame_with_number.store(i, fp) }
+
+total = 0
+frame_with_number.each do |frame, point|
```

Comment:
`frame_with_number`でハッシュを作らなくても、`frames`を配列のまま処理できるんじゃないかなーと思いました。

```suggestion
frames.each.with_index(1) do |frame, point|
```

あと、`do |frame, point|`の部分ですが、`frame`はフレーム番号が入ると思うので、数値が入っているのがわかるような名前がいいですね。
また、`point`は配列なので、`points`のように複数形にするのがベターです。

最後に若干高等テクニックですが、こんな感じでsumメソッドとブロックを使う方法もあります。

```ruby
total = frames.each.with_index(1).sum do |frame, point|
```

Reply:
@JunichiIto 
## do |frame, point|の部分の名前
以下の通り修正いたしました！
- frame → number_of_frameに変更
- point → pointsに変更

### 誤りのご報告
本コミット（6e1796fcdd5fb23bf1295ad36cf96af73541c53b）のコミットメッセージですが、
以下のとおり誤っております＞＜
- 【誤】合計スコア算出時の変数を修正
- 【正】合計スコア算出時のブロックパラメータを修正
　　　　　　　　　　　　　^^^^^^^^^^^^^^

## sumメソッドとブロックを使う方法
こちらもチェリー本で昨日学習した内容でしたので、早速使用するかたちに修正いたしました🙇‍♀️

## framesを配列のまま処理する
- 配列のまま処理する場合のコードを作成、計算結果が問題ないことを確認いたしました。

```
total =
  frames.each.with_index(1).sum do |points, number_of_frame|
    next_frame = frames[number_of_frame]
    after_next_frame = frames[number_of_frame + 1]

    if number_of_frame <= 9 && points[0] == 10 && frames[number_of_frame][0] == 10
      10 + next_frame[0] + after_next_frame[0]
    elsif number_of_frame <= 9 && points[0] == 10
      10 + next_frame.sum
    elsif number_of_frame <= 9 && points.sum == 10
      10 + next_frame[0]
    else
       points.sum
    end
  end
```

### ハッシュで算出した理由
ご教示いただいたとおり、配列のままでも算出できることを上記にて理解できました！
ありがとうございます🙇‍♀️
配列のまま算出するとハッシュにする工程が省けるため全体的にスッキリするのですが、
実際にコードを作成して比較した結果、以下の理由によりハッシュで算出する方法のまま再度プルリクエストを行いました。

- 最終フレームでの3投目は除き、インデックス（ブロックパラメータのnumber_of_frame）をフレーム数に合わせたい
- 上記を実現するとブロックパラメータのnumber_of_frameと、framesの添え字がずれてしまう（添え字は0から始まるため）
  - => 後のフレームを表すコードが以下のようになってしまう
```
#  1つ後ろのフレームを表す変数
next_frame = frames[number_of_frame]

#  2つ後ろのフレームを表す変数
after_next_frame = frames[number_of_frame + 1]
```
- ハッシュを使うと後ろのフレームを表す際もkey（=フレーム数）を使用できるため、個人的にはわかりやすいと感じました🙇‍♀️

Reply:
すいません、ハッシュで算出した理由がちょっとつかめなかったんですが、もしかするとこんなふうにすると配列でも問題なかったりしませんか？

```ruby
frames.each_with_index.sum do |points, i|
  next_frame = frames[i + 1]
  after_next_frame = frames[i + 2]
  number_of_frame = i + 1

  if number_of_frame <= 9 && points[0] == 10 && frames[number_of_frame][0] == 10
    10 + next_frame[0] + after_next_frame[0]
```

基本的に「キーが数値、かつ連番になっているハッシュ」というのはハッシュではなく配列として定義すべきだと思います。
なぜなら、ハッシュよりも配列の方が目的にあったデータ構造だからです。
もちろん、ハッシュで配列のようなデータ構造を作っても、配列と同じようにプログラムは動作しますが、ハッシュで配列のようなことをやるなら、最初から配列を使った方が良いです。
また、若干脱線しますが、プログラミング言語によってはハッシュは要素の並び順を保証しないものがあります。Rubyのハッシュは要素を追加した順番を保持してくれますが、これはプログラミング言語の中では結構例外的です（たしか）。こういう理由もあって、ハッシュで配列のようなことをやるのには「うーん」という気がします。

というわけで、このロジックは配列を使ってみてほしいですー。

Reply:
>もしかするとこんなふうにすると配列でも問題なかったりしませんか？

ご提示いただいたコードで問題ありませんでした...！
ハッシュで算出した理由の箇所、上手くお伝えができず申し訳ありません。
にもかかわらず、意図を汲んでくださりありがとうございます！

頂戴した内容を反映し、配列で算出するよう修正いたします🙇‍♀️

一度each_with_indexで配列に0から開始するインデックスを付ける（frames.each_with_index）
→インデックスに+1することで（number_of_frame = i + 1）配列のインデックスをフレーム数と＝にすることができるのですね！

またハッシュよりも配列として定義すべきとの点承知いたしました！
理由についても詳細にご教示いただきありがとうございます🙇‍♀️

Reply:
ハッシュよりも配列として定義すべきとの点承知いたしました！
理由についても詳細にご教示いただきありがとうございます🙇‍♀️
配列で算出する方法に修正いたします！

#### ハッシュで算出した理由について
ハッシュで算出した理由の箇所、上手くお伝えができず申し訳ありません。
にもかかわらず、コードのご提示までいただきありがとうございます！

`if number_of_frame <= 9 && points[0] == 10 && frames[number_of_frame][0] == 10`

配列だと、number_of_frame （= 現在処理を行っているインデックス=フレーム数）と、
次のフレームを指定する際の添え字（ frames[**number_of_frame**]）が同じ数（number_of_frame）であることになんとなく違和感があり、ハッシュを使用したいと考えてしまっておりました。

※ハッシュだと以下のようになるので、
`if frame <= 9 && point[0] == 10 && frame_with_number[frame + 1][0] == 10`
frame（= 現在処理を行っているインデックス= フレーム数）に対して、
次のフレームを指定する際のインデックス（ frame_with_number[**frame + 1**]）が`frame+1`になっている点が感覚的にしっくりくると感じていました。
（+1で「次のフレーム感」が表せているな〜という感覚です）

ですが、このあたりは個人的な感覚且つ拘るべきところではなく、
ご教示いただいたハッシュと配列の特徴を優先すべきと判断いたしました！
長々と申し訳ありませんでした...🙇‍♀️